File: package/python/src/life_game.py
from typing import List


def next_generation(board: List[List[int]]) -> List[List[int]]:
    """
    Calculate the next generation of Conway's Game of Life.

    Args:
        board: Current board state (m x n grid of 0s and 1s)

    Returns:
        Next generation board state
    """
    if not board or not board[0]:
        return []

    m, n = len(board), len(board[0])
    next_board = [[0] * n for _ in range(m)]

    directions = [(-1, -1), (-1, 0), (-1, 1), (0, -1), (0, 1), (1, -1), (1, 0), (1, 1)]

    for i in range(m):
        for j in range(n):
            live_neighbors = 0

            for di, dj in directions:
                ni, nj = i + di, j + dj
                if 0 <= ni < m and 0 <= nj < n and board[ni][nj] == 1:
                    live_neighbors += 1

            if board[i][j] == 1:
                if live_neighbors == 2 or live_neighbors == 3:
                    next_board[i][j] = 1
                else:
                    next_board[i][j] = 0
            else:
                if live_neighbors == 3:
                    next_board[i][j] = 1
                else:
                    next_board[i][j] = 0

    return next_board

File: package/typescript/src/lifeGame.ts
/**
 * Calculate the next generation of Conway's Game of Life.
 *
 * @param board - Current board state (m x n grid of 0s and 1s)
 * @returns Next generation board state
 */
export function nextGeneration(board: number[][]): number[][] {
  if (!board || !board[0]) {
    return [];
  }

  const m = board.length;
  const n = board[0].length;
  const nextBoard = Array.from({ length: m }, () => Array(n).fill(0));

  const directions = [[-1, -1], [-1, 0], [-1, 1], [0, -1], [0, 1], [1, -1], [1, 0], [1, 1]];

  for (let i = 0; i < m; i++) {
    for (let j = 0; j < n; j++) {
      let liveNeighbors = 0;

      for (const [di, dj] of directions) {
        const ni = i + di;
        const nj = j + dj;
        if (ni >= 0 && ni < m && nj >= 0 && nj < n && board[ni][nj] === 1) {
          liveNeighbors++;
        }
      }

      if (board[i][j] === 1) {
        if (liveNeighbors === 2 || liveNeighbors === 3) {
          nextBoard[i][j] = 1;
        } else {
          nextBoard[i][j] = 0;
        }
      } else {
        if (liveNeighbors === 3) {
          nextBoard[i][j] = 1;
        } else {
          nextBoard[i][j] = 0;
        }
      }
    }
  }

  return nextBoard;
}
